fink_science/snad/processor.py

```diff
@@ -0,0 +1,183 @@
+        lc.AndersonDarlingNormal(),                    #
+        lc.ReducedChi2(),                              # A2.0.25  Reduced𝜒2
+        lc.Skew(),                                     # A2.0.26  Skew
+        lc.StetsonK(),                                 # A2.0.28  Stetson𝐾
+
+    )
+
+
+# 'lc.Extrator' can not be pickled, and thus needs to be created inside the udf,
+# but we also need the list of names outside the udf
+column_names = create_extractor().names
+columns_count = len(column_names)
+
+
+def extract_features_snad_raw(
+    magpsf,
+    jd,
+    sigmapsf,
+    cfid,
+    oId
+) -> pd.Series:
+    """ Returns many features, extracted from measurment's using light_curve package (https://github.com/light-curve/light-curve-python).
+    Reference - https://arxiv.org/pdf/2012.01419.pdf#section.A1
+
+    Parameters
+    ----------
+    jd: Spark DataFrame Column
+        Julian date
+    magpsf, sigmapsf: Spark DataFrame Columns
+        Magnitude from PSF-fit photometry, and 1-sigma error
+    fid: Spark DataFrame Column
+        Filter IDs (int)
+    oId: Spark DataFrame Column
+        Object IDs (str)
+
+    Returns
+    ----------
+    out: list
+        Returns list of lists. Each "inner" list consists of many light curve features for each filter.
+
+    Examples
+    ---------
+    >>> from fink_science.utilities import concat_col
+    >>> from pyspark.sql import functions as F
+
+    >>> df = spark.read.load(ztf_alert_sample)
+
+    # Required alert columns, concatenated with historical data
+    >>> what = ['magpsf', 'jd', 'sigmapsf', 'fid']
+    >>> prefix = 'c'
+    >>> what_prefix = [prefix + i for i in what]
+    >>> for colname in what:
+    ...    df = concat_col(df, colname, prefix=prefix)
+
+    >>> df = df.withColumn('lc_features', extract_features_snad(*what_prefix, 'objectId'))
+
+    >>> for row in df.take(10):
+    ...    assert len(row['lc_features']) == len(np.unique(row['cfid']))
+    ...    assert len(row['lc_features'][1]) == 26
+    """
+
+    cfid = np.asarray(cfid, "int32")
+    magpsf = np.asarray(magpsf, "float64")
+    jd = np.asarray(jd, "float64")
+    sigmapsf = np.asarray(sigmapsf, "float64")
+
+    extractor = create_extractor()
+
+    passbands = np.unique(cfid)
+
+    # Select only valid measurements (not upper limits)
+    maskNotNone = magpsf == magpsf
+    mask = ~(np.isnan(magpsf) | np.isnan(sigmapsf)) & maskNotNone
+
+    magpsf = magpsf[mask]
+    sigmapsf = sigmapsf[mask]
+    jd = jd[mask]
+    cfid = cfid[mask]
+
+    full_result = {}
+    for passband_id in passbands:
+        passband = cfid == passband_id
+        try:
+            result = extractor(jd[passband], magpsf[passband], sigmapsf[passband], fill_value=np.nan)
+        except ValueError as err:
+            # log if known error, then skip
+            if err.args[0] == "t must be in ascending order":
+                logger.error(f"Unordered jd for {oId} in processor '{__file__}/{extract_features_snad.__name__}'")
```

Comment:
I do not understand this error. The columns should be ordered by`jd` according to the line

```python
df = concat_col(df, colname, prefix=prefix)
```

And if I print one of the array, it is indeed ordered:

```
Unordered jd for ZTF18abckozp in processor '/Users/julien/Documents/workspace/myrepos/fink-science/fink_science/__init__.py/wrapper'
[2459517.7868056 2459517.7868056 2459517.7868056 2459517.8475
 2459517.8475    2459517.8475    2459517.8475    2459521.7640162
 2459521.7640162 2459522.8031366 2459530.8480324 2459531.7856481
 2459531.7856481 2459531.7856481 2459537.8261574 2459537.8261574
 2459537.8261574 2459537.8261574 2459540.7134028 2459540.7134028]
```

I think the code is complaining because there are _identical_ `jd`. Which I agree should not exist! I was looking at the data, and it seems a bug from the ZTF alert system whend doing the association between candidates  and objects. I will inspect and come back to you on this.

Reply:
Got it, thanks. I will leave this check for now.

Reply:
No firm conclusion yet from ZTF -- so let's keep the code as-is.